Quorum/llm/chains/first_deposit_chain.py

```diff
@@ -0,0 +1,93 @@
+from typing import Optional
+from pydantic import BaseModel, Field
+from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
+
+from Quorum.llm.chains.cached_llm import CachedLLM
+from Quorum.llm.jinja_utils import render_prompt
+
+
+class ListingDetails(BaseModel):
+    """
+    A model representing details for listing an asset.
+
+    This class extends BaseModel and contains information necessary for listing
+    operations including asset details, supply amount, and operation indicators.
+
+    Attributes:
+        asset (str): The identifier or symbol of the asset to be listed.
+        supply_seed_amount (Optional[float]): The initial amount to seed the supply with.
+            Defaults to None if not specified.
+        supply_indicator (bool): Flag indicating whether a supply operation is being
+            performed for the asset. True if supply is called, False otherwise.
+        approve_indicator (bool): Flag indicating whether an approval operation is
+            being performed for the asset.
+    """
+    asset_symbol: str = Field(description="The asset symbol to be listed.")
+    asset_address: str = Field(description="The asset address to be listed.")
+    supply_seed_amount: Optional[float] = Field(description="The amount of supply seed.")
+    supply_indicator: bool = Field(
+        description="The indicator for supply being call for the asset. True if supply is being called. False otherwise."
+    )
+    approve_indicator: bool = Field(
+        description="The indicator for approval being call for the asset."
+    )
+
+class ListingArray(BaseModel):
+    """
+    A container class for holding multiple listing details in a structured format.
+
+    This class represents an array of `ListingDetails` objects, each containing specific
+    parameters and configurations for asset listings. It is designed to handle multiple
+    listings within a single proposal.
+
+    Attributes:
+        listings (list[ListingDetails]): A list containing individual listing details
+            for each asset proposed to be listed. Each element in the list is a
+            ListingDetails object specifying the complete configuration for that
+            particular asset listing.
+    """
+    listings: list[ListingDetails] = Field(description="The list of assets to be listed.")
+
+class FirstDepositChain(CachedLLM):
+
+    def __init__(self):
+        super().__init__()
+        structured_llm = self.llm.with_structured_output(ListingArray)
```

Comment:
Are you sure you can't pass just a list of Listing? That seems pretty weird.

Reply:
Also let me guess, this created an "llm client" that will automatically use the `description` of each field in the pydantic models in the context of each prompt to the llm to return whatever data type you put inside for every message?

Reply:
1. This is the usage according to the docs, it makes sense that it needs a pydantic obj, https://python.langchain.com/docs/how_to/structured_output/
2. Not sure how dark magic works but it does.

Reply:
That's cool and very important to know. From the page you linked:
"Beyond just the structure of the Pydantic class, the name of the Pydantic class, the docstring, and the names and provided descriptions of parameters are very important. Most of the time with_structured_output is using a model's function/tool calling API, and you can effectively think of all of this information as being added to the model prompt."

Reply:
This is the easiest and most reliable way to get structured outputs. with_structured_output() is implemented for [models that provide native APIs for structuring outputs](https://python.langchain.com/docs/integrations/chat/), like tool/function calling or JSON mode, and makes use of these capabilities under the hood.

Reply:
https://python.langchain.com/docs/integrations/chat/

Basically its a mix between prompt modification and Inherit support in the LLM tool/ API call that makes it happen